Error -19 while starting MediaRecorder (Camera1 only)

### Describe the bug
Please add a clear description of what the bug is, **and** fill the list below.
- CameraView version: 2.2
- Camera engine used: *camera1
- Reproducible in official demo app: *yes*
- Device / Android version: *Moto G6, API 28*

### To Reproduce
Steps to reproduce the behavior, possibly in the demo app:
1. Select camera 1 in sample app and start recording with front camera

### Expected behavior
Camera1, recording should with front camera.

### XML layout
Part of the XML layout with the CameraView declaration, so we can read its attributes.

```xml
    <com.otaliastudios.cameraview.CameraView
        android:id="@+id/camera"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:layout_gravity="center"
        android:keepScreenOn="true"
        app:cameraMode="video"
        app:cameraEngine="camera1"
        app:cameraHdr="off"
        app:cameraPreview="glSurface"
        app:cameraAudio="on"
        app:cameraFacing="front"
        app:cameraFlash="off"
        app:cameraWhiteBalance="auto"
        app:cameraVideoCodec="deviceDefault"
        app:cameraGestureLongTap="none"
        app:cameraGesturePinch="none"
        app:cameraGestureScrollHorizontal="none"
        app:cameraGestureScrollVertical="none"
        app:cameraGestureTap="autoFocus"
        app:cameraGrid="off"
        app:cameraPlaySounds="true"
        app:cameraUseDeviceOrientation="true"

        />

```

### Screenshots
If applicable, add screenshots to help explain your problem.

### Logs
start: Error while starting media recorder. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.

### APK
Expection #5 in sample app. 
Moto G 6

I am having same issue with latest version 2.3.0. Below are logs from demo app. 

I am using Camera1 with front camera on moto(g6) India version.

`2019-10-03 10:44:25.773 20940-20940/com.otaliastudios.cameraview.demo W/DemoApp: Recording for 5 seconds...
2019-10-03 10:44:25.792 20940-20940/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo scheduling
2019-10-03 10:44:25.793 20940-20974/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo performing. BindState: 2 isTakingVideo: false
2019-10-03 10:44:25.812 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: findDeviceEncoder - type: video/avc encoders: 2
2019-10-03 10:44:25.812 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: Enabled. Found video encoder: OMX.qcom.video.encoder.avc
2019-10-03 10:44:25.813 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: findDeviceEncoder - type: audio/mp4a-latm encoders: 1
2019-10-03 10:44:25.813 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: Enabled. Found audio encoder: OMX.google.aac.encoder
2019-10-03 10:44:25.814 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: getSupportedVideoSize - inputSize: 2304x1728 adjustedSize: 1450x1088
2019-10-03 10:44:25.814 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: getSupportedVideoBitRate - inputRate: 17000000 adjustedRate: 17000000
2019-10-03 10:44:25.814 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: getSupportedAudioBitRate - inputRate: 128000 adjustedRate: 128000
2019-10-03 10:44:25.814 20940-20974/com.otaliastudios.cameraview.demo I/DeviceEncoders: getSupportedVideoFrameRate - inputRate: 30 adjustedRate: 30
2019-10-03 10:44:25.828 20940-20974/com.otaliastudios.cameraview.demo E/MediaRecorder: start failed: -19
2019-10-03 10:44:25.829 20940-20974/com.otaliastudios.cameraview.demo W/FullVideoRecorder: start: Error while starting media recorder. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:206)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-03 10:44:25.829 20940-20974/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchOnVideoRecordingEnd
2019-10-03 10:44:25.830 20940-20974/com.otaliastudios.cameraview.demo E/MediaRecorder: stop called in an invalid state: 0
2019-10-03 10:44:25.831 20940-20974/com.otaliastudios.cameraview.demo W/FullVideoRecorder: stop: Error while closing media recorder. java.lang.IllegalStateException
    java.lang.IllegalStateException
        at android.media.MediaRecorder.stop(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:221)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:212)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-03 10:44:25.835 20940-20974/com.otaliastudios.cameraview.demo E/CameraEngine: onVideoResult result is null: something went wrong. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:206)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-03 10:44:25.836 20940-20974/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchError com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
    com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
        at com.otaliastudios.cameraview.engine.CameraEngine.onVideoResult(CameraEngine.java:1277)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onVideoResult(Camera1Engine.java:398)
        at com.otaliastudios.cameraview.video.VideoRecorder.dispatchResult(VideoRecorder.java:115)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:236)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:212)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
     Caused by: java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:206)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76) 
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356) 
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.os.HandlerThread.run(HandlerThread.java:65) 
2019-10-03 10:44:25.859 20940-20940/com.otaliastudios.cameraview.demo I/DemoApp: Video taken. Processing...
2019-10-03 10:44:25.867 20940-20940/com.otaliastudios.cameraview.demo W/DemoApp: onVideoRecordingEnd!
2019-10-03 10:44:25.886 20940-20940/com.otaliastudios.cameraview.demo W/DemoApp: Got CameraException #5
2019-10-03 10:45:14.781 20940-20940/com.otaliastudios.cameraview.demo I/CameraCallbacks: onDeviceOrientationChanged 270
2019-10-03 10:45:14.781 20940-20940/com.otaliastudios.cameraview.demo I/Angles: Angles changed: sensorOffset: 90 displayOffset: 0 deviceOrientation: 270
2019-10-03 10:45:15.766 20940-20940/com.otaliastudios.cameraview.demo I/CameraCallbacks: onDeviceOrientationChanged 0
2019-10-03 10:45:15.766 20940-20940/com.otaliastudios.cameraview.demo I/Angles: Angles changed: sensorOffset: 90 displayOffset: 0 deviceOrientation: 0
`

@natario1 still same issue , logs as below

2019-10-04 11:40:30.894 13269-13323/com.otaliastudios.cameraview.demo V/Frame: Frame with time 1570169430893 is being released.
2019-10-04 11:40:30.903 13269-13269/com.otaliastudios.cameraview.demo W/DemoApp: Recording for 5 seconds...
2019-10-04 11:40:30.918 13269-13269/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo scheduling
2019-10-04 11:40:30.919 13269-13324/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo performing. BindState: 2 isTakingVideo: false
2019-10-04 11:40:30.960 13269-13324/com.otaliastudios.cameraview.demo E/MediaRecorder: start failed: -19
2019-10-04 11:40:30.961 13269-13324/com.otaliastudios.cameraview.demo W/FullVideoRecorder: start: Error while starting media recorder. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 11:40:30.961 13269-13324/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchOnVideoRecordingEnd
2019-10-04 11:40:30.962 13269-13324/com.otaliastudios.cameraview.demo E/MediaRecorder: stop called in an invalid state: 0
2019-10-04 11:40:30.966 13269-13324/com.otaliastudios.cameraview.demo W/FullVideoRecorder: stop: Error while closing media recorder. java.lang.IllegalStateException
    java.lang.IllegalStateException
        at android.media.MediaRecorder.stop(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:224)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:215)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 11:40:30.969 13269-13324/com.otaliastudios.cameraview.demo E/CameraEngine: onVideoResult result is null: something went wrong. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 11:40:30.972 13269-13324/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchError com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
    com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
        at com.otaliastudios.cameraview.engine.CameraEngine.onVideoResult(CameraEngine.java:1277)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onVideoResult(Camera1Engine.java:398)
        at com.otaliastudios.cameraview.video.VideoRecorder.dispatchResult(VideoRecorder.java:115)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:239)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:215)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
     Caused by: java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76) 
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356) 
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.os.HandlerThread.run(HandlerThread.java:65) 
2019-10-04 11:40:30.974 13269-13324/com.otaliastudios.cameraview.demo V/FrameManager: getFrame for time: 1570169430973 RECYCLING.
2019-10-04 11:40:30.974 13269-13324/com.otaliastudios.cameraview.demo V/CameraCallbacks: dispatchFrame: 1570169430973 processors: 3
2019-10-04 11:40:30.974 13269-13323/com.otaliastudios.cameraview.demo V/CameraCallbacks: dispatchFrame: dispatching 1570169430973 to processors.
2019-10-04 11:40:30.974 13269-13323/com.otaliastudios.cameraview.demo V/Frame: Frame with time 1570169430973 is being released.
2019-10-04 11:40:30.993 13269-13269/com.otaliastudios.cameraview.demo I/DemoApp: Video taken. Processing...
2019-10-04 11:40:31.002 13269-13269/com.otaliastudios.cameraview.demo W/DemoApp: onVideoRecordingEnd!
2019-10-04 11:40:31.002 13269-13269/com.otaliastudios.cameraview.demo W/DemoApp: Got CameraException #5

2019-10-04 18:17:53.074 32254-32399/? D/ClClient: Not sending keepalive.  Current connection state=STOPPED
2019-10-04 18:17:55.808 2044-2044/com.otaliastudios.cameraview.demo W/DemoApp: Recording for 5 seconds...
2019-10-04 18:17:55.824 2044-2044/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo scheduling
2019-10-04 18:17:55.824 2044-2167/com.otaliastudios.cameraview.demo V/CameraEngine: takeVideo performing. BindState: 2 isTakingVideo: false
2019-10-04 18:17:55.827 1284-1922/? W/StagefrightRecorder: stop while neither recording nor paused
2019-10-04 18:17:55.840 1284-1922/? W/StagefrightRecorder: Max file size is not positive: 0 bytes. Disabling file size limit.
2019-10-04 18:17:55.840 1284-1922/? W/StagefrightRecorder: Target file size (0 bytes) is too small to be respected
2019-10-04 18:17:55.841 1284-1922/? W/StagefrightRecorder: Target duration (5000000 us) too short to be respected
2019-10-04 18:17:55.844 1284-1922/? W/StagefrightRecorder: Intended video encoding frame height (2304) is too large and will be set to (2160)
2019-10-04 18:17:55.846 1284-1922/? E/CameraSource: Video dimension (1728x2160) is unsupported
2019-10-04 18:17:55.846 1284-1922/? I/ExtendedCameraSource: Careating : ExtendedCameraSource
2019-10-04 18:17:55.846 1275-2452/? E/Camera2ClientBase: getParameters: attempt to use a locked camera from a different process (old pid 0, new pid 1284)
2019-10-04 18:17:55.847 1275-2452/? E/Camera2ClientBase: setParameters: attempt to use a locked camera from a different process (old pid 0, new pid 1284)
2019-10-04 18:17:55.848 1284-1922/? E/ExtendedUtils: Failed to set parameter - cache-video-buffers
2019-10-04 18:17:55.848 2044-2167/com.otaliastudios.cameraview.demo E/MediaRecorder: start failed: -19
2019-10-04 18:17:55.850 2044-2167/com.otaliastudios.cameraview.demo W/FullVideoRecorder: start: Error while starting media recorder. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 18:17:55.850 2044-2167/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchOnVideoRecordingEnd
2019-10-04 18:17:55.850 2044-2167/com.otaliastudios.cameraview.demo E/MediaRecorder: stop called in an invalid state: 0
2019-10-04 18:17:55.851 2044-2167/com.otaliastudios.cameraview.demo W/FullVideoRecorder: stop: Error while closing media recorder. java.lang.IllegalStateException
    java.lang.IllegalStateException
        at android.media.MediaRecorder.stop(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:224)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:215)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 18:17:55.851 1284-1922/? W/StagefrightRecorder: stop while neither recording nor paused
2019-10-04 18:17:55.852 1275-2452/? E/Camera2ClientBase: setPreviewCallbackFlag: attempt to use a locked camera from a different process (old pid 0, new pid 2044)
2019-10-04 18:17:55.853 2044-2167/com.otaliastudios.cameraview.demo E/CameraEngine: onVideoResult result is null: something went wrong. java.lang.RuntimeException: start failed.
    java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-10-04 18:17:55.854 2044-2167/com.otaliastudios.cameraview.demo I/CameraCallbacks: dispatchError com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
    com.otaliastudios.cameraview.CameraException: java.lang.RuntimeException: start failed.
        at com.otaliastudios.cameraview.engine.CameraEngine.onVideoResult(CameraEngine.java:1277)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onVideoResult(Camera1Engine.java:398)
        at com.otaliastudios.cameraview.video.VideoRecorder.dispatchResult(VideoRecorder.java:115)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStop(FullVideoRecorder.java:239)
        at com.otaliastudios.cameraview.video.VideoRecorder.stop(VideoRecorder.java:90)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:215)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76)
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356)
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
     Caused by: java.lang.RuntimeException: start failed.
        at android.media.MediaRecorder.start(Native Method)
        at com.otaliastudios.cameraview.video.FullVideoRecorder.onStart(FullVideoRecorder.java:209)
        at com.otaliastudios.cameraview.video.VideoRecorder.start(VideoRecorder.java:76) 
        at com.otaliastudios.cameraview.engine.Camera1Engine.onTakeVideo(Camera1Engine.java:356) 
        at com.otaliastudios.cameraview.engine.CameraEngine$23.run(CameraEngine.java:1212) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.os.HandlerThread.run(HandlerThread.java:65) 
2019-10-04 18:17:55.930 2044-2044/com.otaliastudios.cameraview.demo I/DemoApp: Video taken. Processing...
2019-10-04 18:17:55.938 2044-2044/com.otaliastudios.cameraview.demo W/DemoApp: onVideoRecordingEnd!
2019-10-04 18:17:55.938 2044-2044/com.otaliastudios.cameraview.demo W/DemoApp: Got CameraException #5
2019-10-04 18:17:55.960 3381-3833/? I/PlayCommon: [165] ahtb.c(28): Preparing logs for uploading
2019-10-04 18:17:55.980 1255-1255/? D/SurfaceFlinger: duplicate layer name: changing Surface(name=e053226 Toast)/@0xce21e67 - animation-leash to Surface(name=e053226 Toast)/@0xce21e67 - animation-leash#1
2019-10-04 18:17:55.985 3381-3833/? I/PlayCommon: [165] ahtb.c(181): Connecting to server: https://play.googleapis.com/play/log?format=raw&proto_v2=true
    
    --------- beginning of system
2019-10-04 18:17:55.988 2057-3275/? W/NotificationService: Toast already killed. pkg=com.otaliastudios.cameraview.demo callback=android.app.ITransientNotification$Stub$Proxy@6e83a03
2019-10-04 18:17:55.995 3381-3833/? E/PlayCommon: [165] ahtb.c(187): Failed to connect to server: java.net.UnknownHostException: Unable to resolve host "play.googleapis.com": No address associated with hostname
2019-10-04 18:17:56.024 1255-1255/? D/SurfaceFlinger: duplicate layer name: changing Toast to Toast#1
2019-10-04 18:17:56.491 1255-2157/? W/SurfaceFlinger: Attempting to set client state on removed layer: Surface(name=e053226 Toast)/@0xce21e67 - animation-leash#1
2019-10-04 18:17:56.492 1255-2157/? W/SurfaceFlinger: Attempting to destroy on removed layer: Surface(name=e053226 Toast)/@0xce21e67 - animation-leash#1
2019-10-04 18:17:59.377 2057-2142/? W/WindowManager: Unable to start animation, surface is null or no children.
2019-10-04 18:17:59.462 2057-3275/? W/NotificationService: Toast already killed. pkg=com.otaliastudios.cameraview.demo callback=android.app.ITransientNotification$Stub$Proxy@b1470d6
2019-10-04 18:17:59.990 1255-2157/? W/SurfaceFlinger: Attempting to destroy on removed layer: 7e7afb9 Toast#0
2019-10-04 18:18:00.025 32254-3200/? D/EventNotificationJob: Running EventNotificationJob, isDetail=true
2019-10-04 18:18:00.035 32254-3200/? D/ACAccountManager: ACMailAccount with ID 1 not found
2019-10-04 18:18:00.035 32254-3200/? D/EventNotificationJob: Could not find account for ID: 1
2019-10-04 18:18:00.037 32254-3200/? D/ACAccountManager: ACMailAccount with ID 1 not found
2019-10-04 18:18:00.037 32254-3200/? D/EventNotificationJob: Could not find account for ID: 1
2019-10-04 18:18:00.038 32254-3200/? D/ACAccountManager: ACMailAccount with ID 1 not found
2019-10-04 18:18:00.038 32254-3200/? D/EventNotificationJob: Could not find account for ID: 1
2019-10-04 18:18:00.110 32254-3200/? D/EventNotificationJob: processEventList() on [23] events. First cleaning up expired notifications.
2019-10-04 18:18:00.115 32254-3200/? D/EventNotificationJob: Finished cleaning up notifications. Generating new notifications next
2019-10-04 18:18:00.119 32254-3200/? D/outlook.Notifications: No reminder enabled. Skip event: [[32768:00000000-0000-0000-e54e-1f0000000000]]
2019-10-04 18:18:00.119 32254-3200/? D/EventNotificationJob:   finished processing events.
2019-10-04 18:18:03.074 32254-32399/? D/ClClient: Not sending keepalive.  Current connection state=STOPPED
2019-10-04 18:18:06.267 947-1001/? I/ADSPD: Watchdog 2 new value: 0004ae9d Last Exp Value: ffffffff
2019-10-04 18:18:06.269 947-1001/? I/ADSPD: Watchdog 3 new value: 0004ae94 Last Exp Value: ffffffff
2019-10-04 18:18:06.279 32254-32314/? D/HxAuthDelegate: GetAccessToken for 73 00000000-0000-0000-a227-1f0000000000 ExpirationInMs 1570187760875 CurrentTimeInMs 1570193286279 Token expired true. Has token true AppInitialized true. If app has not been initialized, that is very unexpected
2019-10-04 18:18:06.280 32254-32314/? D/HxAuthDelegate: refreshTokenForAccount will kick token refresh for 73 00000000-0000-0000-a227-1f0000000000
2019-10-04 18:18:06.281 32254-3018/? I/outlook.Accounts: Forcing token update for account(s) [32768]
2019-10-04 18:18:06.283 32254-3018/? D/PowerManagerSettingsLogger: isDeviceIdleMode false isIgnoringBatteryOptimizations false isInteractive true isPowerSaveMode false isSustainedPerformanceModeSupported false
2019-10-04 18:18:06.284 32254-3018/? D/outlook.Accounts: O365 accountId: 32768 token expiration time: 1570187760875 ms next token refresh at: 1570197186283 ms isForcedRefresh: true
2019-10-04 18:18:06.284 32254-3018/? D/outlook.Accounts: Adding O365 access token to token refresh scope
2019-10-04 18:18:06.285 32254-3018/? I/ADALUtil: Acquire token for account 32768 with resource https://outlook.office365.com/
2019-10-04 18:18:06.290 32254-3018/? D/outlook.Accounts: ADAL auth correlationId: 9dc4b874-2a5e-421f-a462-8e7dc4ff7367
2019-10-04 18:18:06.308 32254-3018/? I/outlook.Accounts: ADALOAuth2TokenCache: [2019-10-04 12:48:06 - {"thread_id":"3645","correlation_id":"9dc4b874-2a5e-421f-a462-8e7dc4ff7367"}] Context is an Application? [true] Android 28 (null) - null
2019-10-04 18:18:06.318 32254-32395/? W/outlook.Accounts: HttpWebRequest: [2019-10-04 12:48:06 - 9dc4b874-2a5e-421f-a462-8e7dc4ff7367] DEVICE_CONNECTION_IS_NOT_AVAILABLE:Connection is not available to refresh token ver:1.16.2 Android 28 (DEVICE_CONNECTION_IS_NOT_AVAILABLE) - null
2019-10-04 18:18:06.321 32254-3018/? E/outlook.Accounts: Token update for account 32768 failed
    com.microsoft.office.outlook.token.TokenUpdateStrategy$TokenUpdateException: com.microsoft.aad.adal.AuthenticationException: Connection is not available to refresh token
        at com.microsoft.office.outlook.token.AadTokenUpdateStrategy$AadTokenAcquirer.acquireTokenSilentSync(AadTokenUpdateStrategy.java:152)
        at com.microsoft.office.outlook.token.Office365TokenUpdateStrategy$Office365TokenAcquirer.acquireTokenSilentSync(Office365TokenUpdateStrategy.java:244)
        at com.microsoft.office.outlook.token.AbstractTokenUpdateStrategy.updateAccountTokensSilent(AbstractTokenUpdateStrategy.java:77)
        at com.microsoft.office.outlook.job.AccountTokenRefreshJob.runAccountTokenRefreshJobCriticalBlock(AccountTokenRefreshJob.java:315)
        at com.microsoft.office.outlook.job.AccountTokenRefreshJob.runAccountTokenRefreshJob(AccountTokenRefreshJob.java:232)
        at com.microsoft.office.outlook.job.AccountTokenRefreshJob.lambda$runAccountTokenRefreshJob$2(AccountTokenRefreshJob.java:659)
        at com.microsoft.office.outlook.job.-$$Lambda$AccountTokenRefreshJob$wOhpbPksBbhzLCEg0sh1yqyW-R8.call(Unknown Source:6)
        at bolts.Task$4.run(Task.java:318)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: com.microsoft.aad.adal.AuthenticationException: Connection is not available to refresh token
        at com.microsoft.aad.adal.HttpWebRequest.throwIfNetworkNotAvailable(HttpWebRequest.java:195)
        at com.microsoft.aad.adal.Discovery.performInstanceDiscovery(Discovery.java:237)
        at com.microsoft.aad.adal.Discovery.validateAuthority(Discovery.java:142)
        at com.microsoft.aad.adal.AcquireTokenRequest.validateAuthority(AcquireTokenRequest.java:312)
        at com.microsoft.aad.adal.AcquireTokenRequest.performAuthorityValidation(AcquireTokenRequest.java:222)
        at com.microsoft.aad.adal.AcquireTokenRequest.validateAcquireTokenRequest(AcquireTokenRequest.java:190)
        at com.microsoft.aad.adal.AcquireTokenRequest.access$100(AcquireTokenRequest.java:55)
        at com.microsoft.aad.adal.AcquireTokenRequest$1.run(AcquireTokenRequest.java:128)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2019-10-04 18:18:06.322 32254-3018/? I/outlook.Accounts: Updated tokens for 0 accounts (1 attempted)
2019-10-04 18:18:06.345 32254-3203/? E/ExceptionStackTrace: The exception is  android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
    libcore.io.Linux.android_getaddrinfo(Native Method)
    libcore.io.BlockGuardOs.android_getaddrinfo(BlockGuardOs.java:172)
    java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:137)
    java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:105)
    java.net.InetAddress.getAllByName(InetAddress.java:1154)
    com.microsoft.office.outlook.http.HttpSortedDns.lookup(HttpSortedDns.java:22)
    okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:185)
    okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:149)
    okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:84)
    okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:214)
    okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
    okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
    okhttp3.internal.c
2019-10-04 18:18:06.345 32254-3203/? E/ExceptionStackTrace: The exception is  Unable to resolve host "outlook.office365.com": No address associated with hostname
    java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:157)
    java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:105)
    java.net.InetAddress.getAllByName(InetAddress.java:1154)
    com.microsoft.office.outlook.http.HttpSortedDns.lookup(HttpSortedDns.java:22)
    okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:185)
    okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:149)
    okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:84)
    okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:214)
    okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
    okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
    okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
    okhttp3.internal.http.RealInterceptorChain.pro
2019-10-04 18:18:10.467 2057-2057/? W/WindowManager: removeWindowToken: Attempted to remove non-existing token: android.os.Binder@eddc48b
2019-10-04 18:18:11.833 954-2573/? V/LocSvc_HIDL_IzatProvider: [onLocationChanged][71] [HS] =>>>> [HC]
2019-10-04 18:18:11.833 2057-2154/? V/LocSvc_HIDL_IzatProvider_jni: [onLocationChanged][208] [HC] <<= [HS]
2019-10-04 18:18:13.074 32254-32399/? D/ClClient: Not sending keepalive.  Current connection state=STOPPED
2019-10-04 18:18:13.150 2057-2103/? I/ActivityManager: Start proc 3207:in.startv.hotstar/u0a880 for service in.startv.hotstar/androidx.work.impl.background.systemjob.SystemJobService
2019-10-04 18:18:13.171 3207-3207/? E/.startv.hotsta: Not starting debugger since process cannot load the jdwp agent.
2019-10-04 18:18:13.221 3207-3207/? I/.startv.hotsta: The ClassLoaderContext is a special shared library.
2019-10-04 18:18:13.338 3207-3207/? V/FA: Registered activity lifecycle callback
2019-10-04 18:18:13.350 3207-3207/? W/v14: Application name is not set. Call Builder#setApplicationName.
2019-10-04 18:18:13.354 3207-3228/? W/v14: Application name is not set. Call Builder#setApplicationName.
2019-10-04 18:18:13.442 3207-3237/? D/NetworkSecurityConfig: Using Network Security Config from resource network_security_config debugBuild: false
2019-10-04 18:18:13.495 1288-1442/? E/storaged: getDiskStats failed with result NOT_SUPPORTED and size 0
2019-10-04 18:18:13.563 3207-3237/? I/.startv.hotsta: Rejecting re-init on previously-failed class java.lang.Class<fv>: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/iid/InstanceIDListenerService;
2019-10-04 18:18:13.563 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.a(java.net.URL) (:66)
2019-10-04 18:18:13.563 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.c(cj0) (:25)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.util.List com.facebook.GraphRequest.a(cj0) (:43)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at dj0 com.facebook.GraphRequest.b() (:3)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at org.json.JSONObject com.facebook.internal.m.a(java.lang.String) (:75)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void com.facebook.internal.m$a.run() (:9)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.lang.Thread.run() (Thread.java:764)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.iid.InstanceIDListenerService" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/base.apk"],nativeLibraryDirectories=[/data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/lib/arm, /data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/base.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:134)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.a(java.net.URL) (:66)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.c(cj0) (:25)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.util.List com.facebook.GraphRequest.a(cj0) (:43)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at dj0 com.facebook.GraphRequest.b() (:3)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at org.json.JSONObject com.facebook.internal.m.a(java.lang.String) (:75)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void com.facebook.internal.m$a.run() (:9)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.lang.Thread.run() (Thread.java:764)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta: Rejecting re-init on previously-failed class java.lang.Class<fv>: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/iid/InstanceIDListenerService;
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.a(java.net.URL) (:66)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.c(cj0) (:25)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.util.List com.facebook.GraphRequest.a(cj0) (:43)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at dj0 com.facebook.GraphRequest.b() (:3)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at org.json.JSONObject com.facebook.internal.m.a(java.lang.String) (:75)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void com.facebook.internal.m$a.run() (:9)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at void java.lang.Thread.run() (Thread.java:764)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.iid.InstanceIDListenerService" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/base.apk"],nativeLibraryDirectories=[/data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/lib/arm, /data/app/in.startv.hotstar-nS7zDKsVlrOK_atmOvRyzQ==/base.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:134)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
2019-10-04 18:18:13.564 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.a(java.net.URL) (:66)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at java.net.HttpURLConnection com.facebook.GraphRequest.c(cj0) (:25)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at java.util.List com.facebook.GraphRequest.a(cj0) (:43)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at dj0 com.facebook.GraphRequest.b() (:3)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at org.json.JSONObject com.facebook.internal.m.a(java.lang.String) (:75)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at void com.facebook.internal.m$a.run() (:9)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
2019-10-04 18:18:13.565 3207-3237/? I/.startv.hotsta:     at void java.lang.Thread.run() (Thread.java:764)
2019-10-04 18:18:13.572 3207-3237/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_feature_enabled'.
2019-10-04 18:18:13.585 3207-3237/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_max_length_minutes'.
2019-10-04 18:18:13.590 3207-3226/? V/FA: Collection enabled
2019-10-04 18:18:13.592 3207-3226/? V/FA: App package, google app id: in.startv.hotstar, 1:849261369085:android:8648cec101e6ff41
2019-10-04 18:18:13.597 3207-3226/? I/FA: App measurement is starting up, version: 14710
2019-10-04 18:18:13.598 3207-3226/? I/FA: To enable debug logging run: adb shell setprop log.tag.FA VERBOSE
2019-10-04 18:18:13.605 3207-3226/? I/FA: To enable faster debug mode event logging run:
      adb shell setprop debug.firebase.analytics.app in.startv.hotstar
2019-10-04 18:18:13.606 3207-3226/? D/FA: Debug-level message logging enabled
2019-10-04 18:18:13.611 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'trace_flimit_time'.
2019-10-04 18:18:13.612 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'trace_flimit_events'.
2019-10-04 18:18:13.615 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'trace_blimit_time'.
2019-10-04 18:18:13.617 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'trace_blimit_events'.
2019-10-04 18:18:13.617 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'network_flimit_time'.
2019-10-04 18:18:13.617 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'network_flimit_events'.
2019-10-04 18:18:13.618 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'network_blimit_time'.
2019-10-04 18:18:13.618 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'network_blimit_events'.
2019-10-04 18:18:13.625 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_feature_enabled'.
2019-10-04 18:18:13.632 3207-3249/? W/FirebasePerformance: Unable to process the PerfMetric due to missing or invalid values. See earlier log statements for additional information on the specific missing/invalid values.
2019-10-04 18:18:13.666 3207-3226/? V/FA: Connecting to remote service
2019-10-04 18:18:13.676 3207-3226/? V/FA: Connection attempt already in progress
2019-10-04 18:18:13.755 3207-3267/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_max_length_minutes'.
2019-10-04 18:18:13.755 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_feature_enabled'.
2019-10-04 18:18:13.756 3207-3249/? W/FirebasePerformance: Unable to process the PerfMetric due to missing or invalid values. See earlier log statements for additional information on the specific missing/invalid values.
2019-10-04 18:18:13.756 3207-3267/? E/Fabric: Settings request failed.
2019-10-04 18:18:13.773 3207-3207/? W/v14: Application name is not set. Call Builder#setApplicationName.
2019-10-04 18:18:14.008 963-963/? I/vendor.rmt_storage: rmt_storage_connect_cb: clnt_h=0x51 conn_h=0xb4717010
2019-10-04 18:18:14.009 963-963/? I/vendor.rmt_storage: rmt_storage_rw_iovec_cb: /boot/modem_fs1: req_h=0x51 msg_id=3: R/W request received
2019-10-04 18:18:14.009 963-963/? I/vendor.rmt_storage: rw_iovec client->path = /boot/modem_fs1
2019-10-04 18:18:14.009 963-963/? I/vendor.rmt_storage: wakelock acquired: 1, error no: 42
2019-10-04 18:18:14.009 963-1116/? I/vendor.rmt_storage: rmt_storage_client_thread: /boot/modem_fs1: Unblock worker thread (th_id: -1269479056)
2019-10-04 18:18:14.021 3207-3207/? E/LOGGER: Logger properties file not found.. continuing with defaults
2019-10-04 18:18:14.022 3207-3207/? E/LOGGER: logfile created FileNotFoundException
2019-10-04 18:18:14.067 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'ABR_VARIANT'.
2019-10-04 18:18:14.068 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'ABR_VARIANT'.
2019-10-04 18:18:14.068 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'MASTHEAD_AUTOPLAY'.
2019-10-04 18:18:14.069 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'MASTHEAD_AUTOPLAY'.
2019-10-04 18:18:14.069 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'VP9_EXPERIMENT'.
2019-10-04 18:18:14.069 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'VP9_EXPERIMENT'.
2019-10-04 18:18:14.074 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'ABR_VARIANT'.
2019-10-04 18:18:14.075 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'ABR_VARIANT'.
2019-10-04 18:18:14.075 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'MASTHEAD_AUTOPLAY'.
2019-10-04 18:18:14.076 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'MASTHEAD_AUTOPLAY'.
2019-10-04 18:18:14.077 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'VP9_EXPERIMENT'.
2019-10-04 18:18:14.077 3207-3207/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'VP9_EXPERIMENT'.
2019-10-04 18:18:14.079 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_feature_enabled'.
2019-10-04 18:18:14.081 3207-3249/? W/FirebasePerformance: Unable to process the PerfMetric due to missing or invalid values. See earlier log statements for additional information on the specific missing/invalid values.
2019-10-04 18:18:14.088 3207-3322/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_max_length_minutes'.
2019-10-04 18:18:14.154 963-1116/? I/vendor.rmt_storage: rmt_storage_client_thread: /boot/modem_fs1: req_h=0x51 msg_id=3: Bytes written = 1572864
2019-10-04 18:18:14.154 963-1116/? I/vendor.rmt_storage: rmt_storage_client_thread: /boot/modem_fs1: req_h=0x51 msg_id=3: Send response: res=0 err=0
2019-10-04 18:18:14.154 963-1116/? I/vendor.rmt_storage: rmt_storage_client_thread: /boot/modem_fs1: About to block rmt_storage client thread (th_id: -1269479056) wakelock released: 1, error no: 0
2019-10-04 18:18:14.158 963-963/? I/vendor.rmt_storage: rmt_storage_disconnect_cb: clnt_h=0x51 conn_h=0xb4717010
2019-10-04 18:18:14.298 3207-3249/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_feature_enabled'.
2019-10-04 18:18:14.298 3207-3246/? W/FirebaseRemoteConfig: No value of type 'String' exists for parameter key 'sessions_max_length_minutes'.
2019-10-04 18:18:14.301 3207-3249/? W/FirebasePerformance: Unable to process the PerfMetric due to missing or invalid values. See earlier log statements for additional information on the specific missing/invalid values.
2019-10-04 18:18:14.305 3207-3226/? D/FA: Connected to remote service
2019-10-04 18:18:14.306 3207-3226/? V/FA: Processing queued up service tasks: 2
2019-10-04 18:18:14.504 2057-3275/? W/DropBoxManagerService: Dropping: keymaster (50 > 0 bytes)
2019-10-04 18:18:14.507 2057-2101/? W/BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.stats.service.DropBoxEntryAddedReceiver
2019-10-04 18:18:14.507 2057-2101/? W/BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$PersistentTrustedReceiver
2019-10-04 18:18:14.525 3575-3344/? E/AQD:PackageSync: Error
2019-10-04 18:18:19.356 3207-3226/? V/FA: Inactivity, disconnecting from the service
2019-10-04 18:18:23.074 32254-32399/? D/ClClient: Not sending keepalive.  Current connection state=STOPPED

--- a/cameraview/src/androidTest/java/com/otaliastudios/cameraview/CameraOptions1Test.java
+++ b/cameraview/src/androidTest/java/com/otaliastudios/cameraview/CameraOptions1Test.java
@@ -41,7 +41,7 @@ public class CameraOptions1Test extends BaseTest {
 
     @Test
     public void testEmpty() {
-        CameraOptions o = new CameraOptions(mock(Camera.Parameters.class), false);
+        CameraOptions o = new CameraOptions(mock(Camera.Parameters.class), 0, false);
         assertTrue(o.getSupportedPictureAspectRatios().isEmpty());
         assertTrue(o.getSupportedPictureSizes().isEmpty());
         assertTrue(o.getSupportedWhiteBalance().isEmpty());
@@ -71,7 +71,7 @@ public class CameraOptions1Test extends BaseTest {
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedPictureSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<Size> supportedSizes = o.getSupportedPictureSizes();
         assertEquals(supportedSizes.size(), sizes.size());
         for (Camera.Size size : sizes) {
@@ -90,7 +90,7 @@ public class CameraOptions1Test extends BaseTest {
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedPictureSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, true);
+        CameraOptions o = new CameraOptions(params, 0, true);
         Collection<Size> supportedSizes = o.getSupportedPictureSizes();
         assertEquals(supportedSizes.size(), sizes.size());
         for (Camera.Size size : sizes) {
@@ -115,7 +115,7 @@ public class CameraOptions1Test extends BaseTest {
 
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedPictureSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<AspectRatio> supportedRatios = o.getSupportedPictureAspectRatios();
         assertEquals(supportedRatios.size(), expected.size());
         for (AspectRatio ratio : expected) {
@@ -134,7 +134,7 @@ public class CameraOptions1Test extends BaseTest {
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<Size> supportedSizes = o.getSupportedVideoSizes();
         assertEquals(supportedSizes.size(), sizes.size());
         for (Camera.Size size : sizes) {
@@ -155,7 +155,7 @@ public class CameraOptions1Test extends BaseTest {
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(null);
         when(params.getSupportedPreviewSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<Size> supportedSizes = o.getSupportedVideoSizes();
         assertEquals(supportedSizes.size(), sizes.size());
         for (Camera.Size size : sizes) {
@@ -174,7 +174,7 @@ public class CameraOptions1Test extends BaseTest {
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, true);
+        CameraOptions o = new CameraOptions(params, 0, true);
         Collection<Size> supportedSizes = o.getSupportedVideoSizes();
         assertEquals(supportedSizes.size(), sizes.size());
         for (Camera.Size size : sizes) {
@@ -199,7 +199,7 @@ public class CameraOptions1Test extends BaseTest {
 
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(sizes);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<AspectRatio> supportedRatios = o.getSupportedVideoAspectRatios();
         assertEquals(supportedRatios.size(), expected.size());
         for (AspectRatio ratio : expected) {
@@ -215,7 +215,7 @@ public class CameraOptions1Test extends BaseTest {
         when(params.getMaxExposureCompensation()).thenReturn(0);
         when(params.getMinExposureCompensation()).thenReturn(0);
 
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         assertFalse(o.supports(GestureAction.AUTO_FOCUS));
         assertTrue(o.supports(GestureAction.TAKE_PICTURE));
         assertTrue(o.supports(GestureAction.NONE));
@@ -229,7 +229,7 @@ public class CameraOptions1Test extends BaseTest {
     public void testAlwaysSupportedControls() {
         // Grid, VideoQuality, SessionType and Audio are always supported.
         Camera.Parameters params = mock(Camera.Parameters.class);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
 
         Collection<Grid> grids = o.getSupportedControls(Grid.class);
         Collection<VideoCodec> video = o.getSupportedControls(VideoCodec.class);
@@ -250,7 +250,7 @@ public class CameraOptions1Test extends BaseTest {
             supported.add(cameraInfo.facing);
         }
 
-        CameraOptions o = new CameraOptions(mock(Camera.Parameters.class), false);
+        CameraOptions o = new CameraOptions(mock(Camera.Parameters.class), 0, false);
         Camera1Mapper m = Camera1Mapper.get();
         Collection<Facing> s = o.getSupportedControls(Facing.class);
         assertEquals(s.size(), supported.size());
@@ -269,7 +269,7 @@ public class CameraOptions1Test extends BaseTest {
                 Camera.Parameters.WHITE_BALANCE_SHADE // Not supported
         ));
 
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<WhiteBalance> w = o.getSupportedControls(WhiteBalance.class);
         assertEquals(w.size(), 2);
         assertTrue(w.contains(WhiteBalance.AUTO));
@@ -288,7 +288,7 @@ public class CameraOptions1Test extends BaseTest {
                 Camera.Parameters.FLASH_MODE_RED_EYE // Not supported
         ));
 
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<Flash> f = o.getSupportedControls(Flash.class);
         assertEquals(f.size(), 3);
         assertTrue(f.contains(Flash.OFF));
@@ -309,7 +309,7 @@ public class CameraOptions1Test extends BaseTest {
                 Camera.Parameters.SCENE_MODE_FIREWORKS // Not supported
         ));
 
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         Collection<Hdr> h = o.getSupportedControls(Hdr.class);
         assertEquals(h.size(), 2);
         assertTrue(h.contains(Hdr.OFF));
@@ -326,7 +326,7 @@ public class CameraOptions1Test extends BaseTest {
         when(params.isZoomSupported()).thenReturn(true);
         //noinspection ArraysAsListWithZeroOrOneArgument
         when(params.getSupportedFocusModes()).thenReturn(Arrays.asList(Camera.Parameters.FOCUS_MODE_AUTO));
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         assertTrue(o.isZoomSupported());
         assertTrue(o.isAutoFocusSupported());
     }
@@ -337,7 +337,7 @@ public class CameraOptions1Test extends BaseTest {
         when(params.getMaxExposureCompensation()).thenReturn(10);
         when(params.getMinExposureCompensation()).thenReturn(-10);
         when(params.getExposureCompensationStep()).thenReturn(0.5f);
-        CameraOptions o = new CameraOptions(params, false);
+        CameraOptions o = new CameraOptions(params, 0, false);
         assertTrue(o.isExposureCorrectionSupported());
         assertEquals(o.getExposureCorrectionMinValue(), -10f * 0.5f, 0f);
         assertEquals(o.getExposureCorrectionMaxValue(), 10f * 0.5f, 0f);

--- a/cameraview/src/androidTest/java/com/otaliastudios/cameraview/CameraOptions1Test.java
+++ b/cameraview/src/androidTest/java/com/otaliastudios/cameraview/CameraOptions1Test.java
@@ -126,11 +126,13 @@ public class CameraOptions1Test extends BaseTest {
 
     @Test
     public void testVideoSizes() {
+        // VideoSize are capped by CamcorderProfile.QUALITY_HIGH max size.
+        // This can be very small on an emulator, so use very small sizes to not hit that value.
         List<Camera.Size> sizes = Arrays.asList(
-                mockCameraSize(100, 200),
-                mockCameraSize(50, 50),
-                mockCameraSize(1600, 900),
-                mockCameraSize(1000, 2000)
+                mockCameraSize(10, 20),
+                mockCameraSize(5, 5),
+                mockCameraSize(16, 9),
+                mockCameraSize(20, 40)
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(sizes);
@@ -147,10 +149,10 @@ public class CameraOptions1Test extends BaseTest {
     public void testVideoSizesNull() {
         // When videoSizes is null, we take the preview sizes.
         List<Camera.Size> sizes = Arrays.asList(
-                mockCameraSize(100, 200),
-                mockCameraSize(50, 50),
-                mockCameraSize(1600, 900),
-                mockCameraSize(1000, 2000)
+                mockCameraSize(10, 20),
+                mockCameraSize(5, 5),
+                mockCameraSize(16, 9),
+                mockCameraSize(20, 40)
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(null);
@@ -167,10 +169,10 @@ public class CameraOptions1Test extends BaseTest {
     @Test
     public void testVideoSizesFlip() {
         List<Camera.Size> sizes = Arrays.asList(
-                mockCameraSize(100, 200),
-                mockCameraSize(50, 50),
-                mockCameraSize(1600, 900),
-                mockCameraSize(1000, 2000)
+                mockCameraSize(10, 20),
+                mockCameraSize(5, 5),
+                mockCameraSize(16, 9),
+                mockCameraSize(20, 40)
         );
         Camera.Parameters params = mock(Camera.Parameters.class);
         when(params.getSupportedVideoSizes()).thenReturn(sizes);
@@ -186,10 +188,10 @@ public class CameraOptions1Test extends BaseTest {
     @Test
     public void testVideoAspectRatio() {
         List<Camera.Size> sizes = Arrays.asList(
-                mockCameraSize(100, 200),
-                mockCameraSize(50, 50),
-                mockCameraSize(1600, 900),
-                mockCameraSize(1000, 2000)
+                mockCameraSize(10, 20),
+                mockCameraSize(5, 5),
+                mockCameraSize(16, 9),
+                mockCameraSize(20, 40)
         );
 
         Set<AspectRatio> expected = new HashSet<>();

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/CameraOptions.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/CameraOptions.java
@@ -62,7 +62,7 @@ public class CameraOptions {
     private boolean autoFocusSupported;
 
 
-    public CameraOptions(@NonNull Camera.Parameters params, boolean flipSizes) {
+    public CameraOptions(@NonNull Camera.Parameters params, int cameraId, boolean flipSizes) {
         List<String> strings;
         Camera1Mapper mapper = Camera1Mapper.get();
 
@@ -127,23 +127,33 @@ public class CameraOptions {
         }
 
         // Video Sizes
+        // As a safety measure, remove Sizes bigger than CamcorderProfile.highest
+        CamcorderProfile profile = CamcorderProfiles.get(cameraId,
+                new Size(Integer.MAX_VALUE, Integer.MAX_VALUE));
+        Size videoMaxSize = new Size(profile.videoFrameWidth, profile.videoFrameHeight);
         List<Camera.Size> vsizes = params.getSupportedVideoSizes();
         if (vsizes != null) {
             for (Camera.Size size : vsizes) {
-                int width = flipSizes ? size.height : size.width;
-                int height = flipSizes ? size.width : size.height;
-                supportedVideoSizes.add(new Size(width, height));
-                supportedVideoAspectRatio.add(AspectRatio.of(width, height));
+                if (size.width <= videoMaxSize.getWidth()
+                        && size.height <= videoMaxSize.getHeight()) {
+                    int width = flipSizes ? size.height : size.width;
+                    int height = flipSizes ? size.width : size.height;
+                    supportedVideoSizes.add(new Size(width, height));
+                    supportedVideoAspectRatio.add(AspectRatio.of(width, height));
+                }
             }
         } else {
             // StackOverflow threads seems to agree that if getSupportedVideoSizes is null,
             // previews can be used.
             List<Camera.Size> fallback = params.getSupportedPreviewSizes();
             for (Camera.Size size : fallback) {
-                int width = flipSizes ? size.height : size.width;
-                int height = flipSizes ? size.width : size.height;
-                supportedVideoSizes.add(new Size(width, height));
-                supportedVideoAspectRatio.add(AspectRatio.of(width, height));
+                if (size.width <= videoMaxSize.getWidth()
+                        && size.height <= videoMaxSize.getHeight()) {
+                    int width = flipSizes ? size.height : size.width;
+                    int height = flipSizes ? size.width : size.height;
+                    supportedVideoSizes.add(new Size(width, height));
+                    supportedVideoAspectRatio.add(AspectRatio.of(width, height));
+                }
             }
         }
     }

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/engine/Camera1Engine.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/engine/Camera1Engine.java
@@ -154,8 +154,8 @@ public class Camera1Engine extends CameraEngine implements
         // Set parameters that might have been set before the camera was opened.
         LOG.i("onStartEngine:", "Applying default parameters.");
         Camera.Parameters params = mCamera.getParameters();
-        mCameraOptions = new CameraOptions(params, getAngles()
-                .flip(Reference.SENSOR, Reference.VIEW));
+        mCameraOptions = new CameraOptions(params, mCameraId,
+                getAngles().flip(Reference.SENSOR, Reference.VIEW));
         applyAllParameters(params);
         mCamera.setParameters(params);
         mCamera.setDisplayOrientation(getAngles().offset(Reference.SENSOR, Reference.VIEW,

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/internal/DeviceEncoders.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/internal/DeviceEncoders.java
@@ -221,22 +221,28 @@ public class DeviceEncoders {
         int width = size.getWidth();
         int height = size.getHeight();
         double aspect = (double) width / height;
+        LOG.i("getSupportedVideoSize - started. width:", width, "height:", height);
 
         // If width is too large, scale down, but keep aspect ratio.
         if (mVideoCapabilities.getSupportedWidths().getUpper() < width) {
             width = mVideoCapabilities.getSupportedWidths().getUpper();
             height = (int) Math.round(width / aspect);
+            LOG.i("getSupportedVideoSize - exceeds maxWidth! width:", width,
+                    "height:", height);
         }
 
         // If height is too large, scale down, but keep aspect ratio.
         if (mVideoCapabilities.getSupportedHeights().getUpper() < height) {
             height = mVideoCapabilities.getSupportedHeights().getUpper();
             width = (int) Math.round(aspect * height);
+            LOG.i("getSupportedVideoSize - exceeds maxHeight! width:", width,
+                    "height:", height);
         }
 
         // Adjust the alignment.
         while (width % mVideoCapabilities.getWidthAlignment() != 0) width--;
         while (height % mVideoCapabilities.getHeightAlignment() != 0) height--;
+        LOG.i("getSupportedVideoSize - aligned. width:", width, "height:", height);
 
         // It's still possible that we're BELOW the lower.
         if (!mVideoCapabilities.getSupportedWidths().contains(width)) {
@@ -256,9 +262,7 @@ public class DeviceEncoders {
                     " Might be an aspect ratio issue." +
                     " Desired size:" + new Size(width, height));
         }
-        Size adjusted = new Size(width, height);
-        LOG.i("getSupportedVideoSize -", "inputSize:", size, "adjustedSize:", adjusted);
-        return adjusted;
+        return new Size(width, height);
     }
 
     /**

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/internal/utils/CamcorderProfiles.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/internal/utils/CamcorderProfiles.java
@@ -4,6 +4,7 @@ import android.annotation.SuppressLint;
 import android.media.CamcorderProfile;
 import android.os.Build;
 
+import com.otaliastudios.cameraview.CameraLogger;
 import com.otaliastudios.cameraview.size.Size;
 
 import java.util.ArrayList;
@@ -19,6 +20,9 @@ import androidx.annotation.NonNull;
  * Wraps the {@link android.media.CamcorderProfile} static utilities.
  */
 public class CamcorderProfiles {
+
+    private static final String TAG = CamcorderProfiles.class.getSimpleName();
+    private static final CameraLogger LOG = CameraLogger.create(TAG);
 
     @SuppressLint("UseSparseArrays")
     private static Map<Size, Integer> sizeToProfileMap = new HashMap<>();
@@ -53,6 +57,7 @@ public class CamcorderProfiles {
             int camera1Id = Integer.parseInt(cameraId);
             return get(camera1Id, targetSize);
         } catch (NumberFormatException e) {
+            LOG.w("NumberFormatException for Camera2 id:", cameraId);
             return CamcorderProfile.get(CamcorderProfile.QUALITY_LOW);
         }
     }

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/video/Full1VideoRecorder.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/video/Full1VideoRecorder.java
@@ -1,6 +1,7 @@
 package com.otaliastudios.cameraview.video;
 
 import android.hardware.Camera;
+import android.media.CamcorderProfile;
 import android.media.MediaRecorder;
 import android.util.Log;
 
@@ -34,15 +35,18 @@ public class Full1VideoRecorder extends FullVideoRecorder {
     }
 
     @Override
-    protected boolean onPrepareMediaRecorder(@NonNull VideoResult.Stub stub,
-                                             @NonNull MediaRecorder mediaRecorder) {
+    protected void applyVideoSource(@NonNull VideoResult.Stub stub,
+                                    @NonNull MediaRecorder mediaRecorder) {
         mediaRecorder.setCamera(mCamera);
         mediaRecorder.setVideoSource(MediaRecorder.VideoSource.CAMERA);
+    }
 
+    @NonNull
+    @Override
+    protected CamcorderProfile getCamcorderProfile(@NonNull VideoResult.Stub stub) {
         // Get a profile of quality compatible with the chosen size.
         Size size = stub.rotation % 180 != 0 ? stub.size.flip() : stub.size;
-        mProfile = CamcorderProfiles.get(mCameraId, size);
-        return super.onPrepareMediaRecorder(stub, mediaRecorder);
+        return CamcorderProfiles.get(mCameraId, size);
     }
 
     @Override

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/video/Full2VideoRecorder.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/video/Full2VideoRecorder.java
@@ -4,6 +4,7 @@ import android.annotation.SuppressLint;
 import android.hardware.camera2.CaptureRequest;
 import android.hardware.camera2.CaptureResult;
 import android.hardware.camera2.TotalCaptureResult;
+import android.media.CamcorderProfile;
 import android.media.MediaRecorder;
 import android.os.Build;
 import android.view.Surface;
@@ -69,19 +70,22 @@ public class Full2VideoRecorder extends FullVideoRecorder {
         action.start(mHolder);
     }
 
-    @SuppressLint("NewApi")
     @Override
-    protected boolean onPrepareMediaRecorder(@NonNull VideoResult.Stub stub,
-                                             @NonNull MediaRecorder mediaRecorder) {
+    protected void applyVideoSource(@NonNull VideoResult.Stub stub,
+                                    @NonNull MediaRecorder mediaRecorder) {
         mediaRecorder.setVideoSource(MediaRecorder.VideoSource.SURFACE);
-        Size size = stub.rotation % 180 != 0 ? stub.size.flip() : stub.size;
-        mProfile = CamcorderProfiles.get(mCameraId, size);
+    }
+
+    @NonNull
+    @Override
+    protected CamcorderProfile getCamcorderProfile(@NonNull VideoResult.Stub stub) {
         // This was an option: get the surface from outside this class, using
         // MediaCodec.createPersistentInputSurface(). But it doesn't really help since the
         // Camera2 engine refuses a surface that has not been configured, so even with that trick
         // we would have to attach the surface to this recorder before creating the CameraSession.
         // mediaRecorder.setInputSurface(mInputSurface);
-        return super.onPrepareMediaRecorder(stub, mediaRecorder);
+        Size size = stub.rotation % 180 != 0 ? stub.size.flip() : stub.size;
+        return CamcorderProfiles.get(mCameraId, size);
     }
 
     /**

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/video/FullVideoRecorder.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/video/FullVideoRecorder.java
@@ -8,6 +8,7 @@ import com.otaliastudios.cameraview.VideoResult;
 import com.otaliastudios.cameraview.controls.Audio;
 import com.otaliastudios.cameraview.controls.VideoCodec;
 import com.otaliastudios.cameraview.internal.DeviceEncoders;
+import com.otaliastudios.cameraview.internal.utils.CamcorderProfiles;
 import com.otaliastudios.cameraview.size.Size;
 
 import androidx.annotation.NonNull;
@@ -17,10 +18,9 @@ import androidx.annotation.Nullable;
  * A {@link VideoRecorder} that uses {@link android.media.MediaRecorder} APIs.
  *
  * When started, the media recorder will be prepared in
- * {@link #onPrepareMediaRecorder(VideoResult.Stub, MediaRecorder)}.
- * Subclasses should override this method and, before calling super(), do two things:
- * - set the media recorder VideoSource
- * - define {@link #mProfile}
+ * {@link #prepareMediaRecorder(VideoResult.Stub)}. This will call two abstract methods:
+ * - {@link #getCamcorderProfile(VideoResult.Stub)}
+ * - {@link #applyVideoSource(VideoResult.Stub, MediaRecorder)}
  *
  * Subclasses can also call {@link #prepareMediaRecorder(VideoResult.Stub)} before start happens,
  * in which case it will not be prepared twice. This can be used for example to test some
@@ -32,7 +32,7 @@ public abstract class FullVideoRecorder extends VideoRecorder {
     private static final CameraLogger LOG = CameraLogger.create(TAG);
 
     @SuppressWarnings("WeakerAccess") protected MediaRecorder mMediaRecorder;
-    @SuppressWarnings("WeakerAccess") protected CamcorderProfile mProfile;
+    private CamcorderProfile mProfile;
     private boolean mMediaRecorderPrepared;
 
 
@@ -40,59 +40,134 @@ public abstract class FullVideoRecorder extends VideoRecorder {
         super(listener);
     }
 
-    @SuppressWarnings({"WeakerAccess", "UnusedReturnValue"})
-    protected boolean prepareMediaRecorder(@NonNull VideoResult.Stub stub) {
+    /**
+     * Subclasses should return an appropriate CamcorderProfile.
+     * This could be taken from the {@link CamcorderProfiles} utility class based on the
+     * stub declared size, for instance.
+     *
+     * @param stub the stub
+     * @return the profile
+     */
+    @NonNull
+    protected abstract CamcorderProfile getCamcorderProfile(@NonNull VideoResult.Stub stub);
+
+    /**
+     * Subclasses should apply a video source to the given recorder.
+     *
+     * @param stub the stub
+     * @param mediaRecorder the recorder
+     */
+    protected abstract void applyVideoSource(@NonNull VideoResult.Stub stub,
+                                             @NonNull MediaRecorder mediaRecorder);
+
+    @SuppressWarnings("WeakerAccess")
+    protected final boolean prepareMediaRecorder(@NonNull VideoResult.Stub stub) {
         if (mMediaRecorderPrepared) return true;
-        return onPrepareMediaRecorder(stub, new MediaRecorder());
+        // We kind of trust the stub size at this point. It's coming from CameraOptions sizes
+        // and it's clipped to be less than CamcorderProfile's highest available profile.
+        // However, we still can't trust the developer parameters (e.g. bit rates), and even
+        // without them, the camera declared sizes can cause crashes in MediaRecorder (#467, #602).
+        // A possible solution was to prepare without checking DeviceEncoders first, and should it
+        // fail, prepare again checking them. However, when parameters are wrong, MediaRecorder
+        // fails on start() instead of prepare() (start failed -19), so this wouldn't be effective.
+        return prepareMediaRecorder(stub, true);
     }
 
-    protected boolean onPrepareMediaRecorder(@NonNull VideoResult.Stub stub,
-                                             @NonNull MediaRecorder mediaRecorder) {
-        mMediaRecorder = mediaRecorder;
+    @SuppressWarnings("SameParameterValue")
+    private boolean prepareMediaRecorder(@NonNull VideoResult.Stub stub,
+                                         boolean applyEncodersConstraints) {
+        // 1. Create reference and ask for the CamcorderProfile
+        mMediaRecorder = new MediaRecorder();
+        mProfile = getCamcorderProfile(stub);
+
+        // 2. Set the video and audio sources.
+        applyVideoSource(stub, mMediaRecorder);
         boolean hasAudio = stub.audio == Audio.ON
                 || stub.audio == Audio.MONO
                 || stub.audio == Audio.STEREO;
         if (hasAudio) {
             mMediaRecorder.setAudioSource(MediaRecorder.AudioSource.DEFAULT);
         }
-        mMediaRecorder.setOutputFormat(mProfile.fileFormat);
 
-        // Get the audio mime type
-        // https://android.googlesource.com/platform/frameworks/av/+/master/media/libmediaplayerservice/StagefrightRecorder.cpp#1096
-        // https://github.com/MrAlex94/Waterfox-Old/blob/master/media/libstagefright/frameworks/av/media/libstagefright/MediaDefs.cpp
-        String audioType;
-        switch (mProfile.audioCodec) {
-            case MediaRecorder.AudioEncoder.AMR_NB: audioType = "audio/3gpp"; break;
-            case MediaRecorder.AudioEncoder.AMR_WB: audioType = "audio/amr-wb"; break;
-            case MediaRecorder.AudioEncoder.AAC:
-            case MediaRecorder.AudioEncoder.HE_AAC:
-            case MediaRecorder.AudioEncoder.AAC_ELD: audioType = "audio/mp4a-latm"; break;
-            case MediaRecorder.AudioEncoder.VORBIS: audioType = "audio/vorbis"; break;
-            case MediaRecorder.AudioEncoder.DEFAULT:
-            default: audioType = "audio/3gpp";
-        }
-
-        // Get the video mime type
-        // https://android.googlesource.com/platform/frameworks/av/+/master/media/libmediaplayerservice/StagefrightRecorder.cpp#1650
-        // https://github.com/MrAlex94/Waterfox-Old/blob/master/media/libstagefright/frameworks/av/media/libstagefright/MediaDefs.cpp
-        String videoType;
+        // 3. Set the output format. Before, change the profile data if the user
+        // has specified a specific codec.
         if (stub.videoCodec == VideoCodec.H_264) {
             mProfile.videoCodec = MediaRecorder.VideoEncoder.H264;
-        }
-        if (stub.videoCodec == VideoCodec.H_263) {
+            mProfile.fileFormat = MediaRecorder.OutputFormat.MPEG_4;
+        } else if (stub.videoCodec == VideoCodec.H_263) {
             mProfile.videoCodec = MediaRecorder.VideoEncoder.H263;
+            mProfile.fileFormat = MediaRecorder.OutputFormat.MPEG_4; // should work
         }
-        switch (mProfile.videoCodec) {
-            case MediaRecorder.VideoEncoder.H263: videoType = "video/3gpp"; break;
-            case MediaRecorder.VideoEncoder.H264: videoType = "video/avc"; break;
-            case MediaRecorder.VideoEncoder.MPEG_4_SP: videoType = "video/mp4v-es"; break;
-            case MediaRecorder.VideoEncoder.VP8: videoType = "video/x-vnd.on2.vp8"; break;
-            case MediaRecorder.VideoEncoder.HEVC: videoType = "video/hevc"; break;
-            case MediaRecorder.VideoEncoder.DEFAULT:
-            default: videoType = "video/avc";
+        mMediaRecorder.setOutputFormat(mProfile.fileFormat);
+
+        // 4. Update the VideoResult stub with DeviceEncoders constraints
+        if (applyEncodersConstraints) {
+            // A. Get the audio mime type
+            // https://android.googlesource.com/platform/frameworks/av/+/master/media/libmediaplayerservice/StagefrightRecorder.cpp#1096
+            // https://github.com/MrAlex94/Waterfox-Old/blob/master/media/libstagefright/frameworks/av/media/libstagefright/MediaDefs.cpp
+            String audioType;
+            switch (mProfile.audioCodec) {
+                case MediaRecorder.AudioEncoder.AMR_NB: audioType = "audio/3gpp"; break;
+                case MediaRecorder.AudioEncoder.AMR_WB: audioType = "audio/amr-wb"; break;
+                case MediaRecorder.AudioEncoder.AAC:
+                case MediaRecorder.AudioEncoder.HE_AAC:
+                case MediaRecorder.AudioEncoder.AAC_ELD: audioType = "audio/mp4a-latm"; break;
+                case MediaRecorder.AudioEncoder.VORBIS: audioType = "audio/vorbis"; break;
+                case MediaRecorder.AudioEncoder.DEFAULT:
+                default: audioType = "audio/3gpp";
+            }
+            // B. Get the video mime type
+            // https://android.googlesource.com/platform/frameworks/av/+/master/media/libmediaplayerservice/StagefrightRecorder.cpp#1650
+            // https://github.com/MrAlex94/Waterfox-Old/blob/master/media/libstagefright/frameworks/av/media/libstagefright/MediaDefs.cpp
+            String videoType;
+            switch (mProfile.videoCodec) {
+                case MediaRecorder.VideoEncoder.H263: videoType = "video/3gpp"; break;
+                case MediaRecorder.VideoEncoder.H264: videoType = "video/avc"; break;
+                case MediaRecorder.VideoEncoder.MPEG_4_SP: videoType = "video/mp4v-es"; break;
+                case MediaRecorder.VideoEncoder.VP8: videoType = "video/x-vnd.on2.vp8"; break;
+                case MediaRecorder.VideoEncoder.HEVC: videoType = "video/hevc"; break;
+                case MediaRecorder.VideoEncoder.DEFAULT:
+                default: videoType = "video/avc";
+            }
+            // C. Check DeviceEncoders support
+            boolean flip = stub.rotation % 180 != 0;
+            if (flip) stub.size = stub.size.flip();
+            Size newVideoSize = null;
+            int newVideoBitRate = 0;
+            int newAudioBitRate = 0;
+            int newVideoFrameRate = 0;
+            int videoEncoderOffset = 0;
+            int audioEncoderOffset = 0;
+            boolean encodersFound = false;
+            while (!encodersFound) {
+                LOG.i("prepareMediaRecorder:", "Checking DeviceEncoders...",
+                        "videoOffset:", videoEncoderOffset,
+                        "audioOffset:", audioEncoderOffset);
+                DeviceEncoders encoders = new DeviceEncoders(DeviceEncoders.MODE_RESPECT_ORDER,
+                        videoType, audioType, videoEncoderOffset, audioEncoderOffset);
+                try {
+                    newVideoSize = encoders.getSupportedVideoSize(stub.size);
+                    newVideoBitRate = encoders.getSupportedVideoBitRate(stub.videoBitRate);
+                    newAudioBitRate = encoders.getSupportedAudioBitRate(stub.audioBitRate);
+                    newVideoFrameRate = encoders.getSupportedVideoFrameRate(newVideoSize,
+                            stub.videoFrameRate);
+                    encodersFound = true;
+                } catch (DeviceEncoders.VideoException videoException) {
+                    videoEncoderOffset++;
+                } catch (DeviceEncoders.AudioException audioException) {
+                    audioEncoderOffset++;
+                }
+            }
+            // D. Apply results
+            stub.size = newVideoSize;
+            stub.videoBitRate = newVideoBitRate;
+            stub.audioBitRate = newAudioBitRate;
+            stub.videoFrameRate = newVideoFrameRate;
+            if (flip) stub.size = stub.size.flip();
         }
 
-        // Merge stub and profile
+        // 5. Update the VideoResult stub with information from the profile, if the
+        // stub values are absent or incomplete
         stub.videoFrameRate = stub.videoFrameRate > 0 ? stub.videoFrameRate
                 : mProfile.videoFrameRate;
         stub.videoBitRate = stub.videoBitRate > 0 ? stub.videoBitRate : mProfile.videoBitRate;
@@ -100,39 +175,8 @@ public abstract class FullVideoRecorder extends VideoRecorder {
             stub.audioBitRate = stub.audioBitRate > 0 ? stub.audioBitRate : mProfile.audioBitRate;
         }
 
-        // Check DeviceEncoders support
+        // 6A. Configure MediaRecorder from stub and from profile (video)
         boolean flip = stub.rotation % 180 != 0;
-        if (flip) stub.size = stub.size.flip();
-        Size newVideoSize = null;
-        int newVideoBitRate = 0;
-        int newAudioBitRate = 0;
-        int newVideoFrameRate = 0;
-        int videoEncoderOffset = 0;
-        int audioEncoderOffset = 0;
-        boolean encodersFound = false;
-        while (!encodersFound) {
-            DeviceEncoders encoders = new DeviceEncoders(DeviceEncoders.MODE_RESPECT_ORDER,
-                    videoType, audioType, videoEncoderOffset, audioEncoderOffset);
-            try {
-                newVideoSize = encoders.getSupportedVideoSize(stub.size);
-                newVideoBitRate = encoders.getSupportedVideoBitRate(stub.videoBitRate);
-                newAudioBitRate = encoders.getSupportedAudioBitRate(stub.audioBitRate);
-                newVideoFrameRate = encoders.getSupportedVideoFrameRate(newVideoSize,
-                        stub.videoFrameRate);
-                encodersFound = true;
-            } catch (DeviceEncoders.VideoException videoException) {
-                videoEncoderOffset++;
-            } catch (DeviceEncoders.AudioException audioException) {
-                audioEncoderOffset++;
-            }
-        }
-        stub.size = newVideoSize;
-        stub.videoBitRate = newVideoBitRate;
-        stub.audioBitRate = newAudioBitRate;
-        stub.videoFrameRate = newVideoFrameRate;
-        if (flip) stub.size = stub.size.flip();
-
-        // Set video params
         mMediaRecorder.setVideoSize(
                 flip ? stub.size.getHeight() : stub.size.getWidth(),
                 flip ? stub.size.getWidth() : stub.size.getHeight());
@@ -140,7 +184,7 @@ public abstract class FullVideoRecorder extends VideoRecorder {
         mMediaRecorder.setVideoEncoder(mProfile.videoCodec);
         mMediaRecorder.setVideoEncodingBitRate(stub.videoBitRate);
 
-        // Set audio params
+        // 6B. Configure MediaRecorder from stub and from profile (audio)
         if (hasAudio) {
             if (stub.audio == Audio.ON) {
                 mMediaRecorder.setAudioChannels(mProfile.audioChannels);
@@ -154,7 +198,7 @@ public abstract class FullVideoRecorder extends VideoRecorder {
             mMediaRecorder.setAudioEncodingBitRate(stub.audioBitRate);
         }
 
-        // Set other params
+        // 7. Set other params
         if (stub.location != null) {
             mMediaRecorder.setLocation(
                     (float) stub.location.getLatitude(),
@@ -180,7 +224,7 @@ public abstract class FullVideoRecorder extends VideoRecorder {
             }
         });
 
-        // Prepare the Recorder
+        // 8. Prepare the Recorder
         try {
             mMediaRecorder.prepare();
             mMediaRecorderPrepared = true;

--- a/cameraview/src/main/java/com/otaliastudios/cameraview/video/FullVideoRecorder.java
+++ b/cameraview/src/main/java/com/otaliastudios/cameraview/video/FullVideoRecorder.java
@@ -100,7 +100,13 @@ public abstract class FullVideoRecorder extends VideoRecorder {
         }
         mMediaRecorder.setOutputFormat(mProfile.fileFormat);
 
-        // 4. Update the VideoResult stub with DeviceEncoders constraints
+        // 4. Update the VideoResult stub with information from the profile, if the
+        // stub values are absent or incomplete
+        if (stub.videoFrameRate <= 0) stub.videoFrameRate = mProfile.videoFrameRate;
+        if (stub.videoBitRate <= 0) stub.videoBitRate = mProfile.videoBitRate;
+        if (stub.audioBitRate <= 0 && hasAudio) stub.audioBitRate = mProfile.audioBitRate;
+
+        // 5. Update the VideoResult stub with DeviceEncoders constraints
         if (applyEncodersConstraints) {
             // A. Get the audio mime type
             // https://android.googlesource.com/platform/frameworks/av/+/master/media/libmediaplayerservice/StagefrightRecorder.cpp#1096
@@ -164,15 +170,6 @@ public abstract class FullVideoRecorder extends VideoRecorder {
             stub.audioBitRate = newAudioBitRate;
             stub.videoFrameRate = newVideoFrameRate;
             if (flip) stub.size = stub.size.flip();
-        }
-
-        // 5. Update the VideoResult stub with information from the profile, if the
-        // stub values are absent or incomplete
-        stub.videoFrameRate = stub.videoFrameRate > 0 ? stub.videoFrameRate
-                : mProfile.videoFrameRate;
-        stub.videoBitRate = stub.videoBitRate > 0 ? stub.videoBitRate : mProfile.videoBitRate;
-        if (hasAudio) {
-            stub.audioBitRate = stub.audioBitRate > 0 ? stub.audioBitRate : mProfile.audioBitRate;
         }
 
         // 6A. Configure MediaRecorder from stub and from profile (video)
@@ -264,13 +261,15 @@ public abstract class FullVideoRecorder extends VideoRecorder {
             try {
                 mMediaRecorder.stop();
             } catch (Exception e) {
-                LOG.w("stop:", "Error while closing media recorder.", e);
                 // This can happen if stopVideo() is called right after takeVideo()
                 // (in which case we don't care). Or when prepare()/start() have failed for
                 // some reason and we are not allowed to call stop.
                 // Make sure we don't override the error if one exists already.
                 mResult = null;
-                if (mError == null) mError = e;
+                if (mError == null) {
+                    LOG.w("stop:", "Error while closing media recorder.", e);
+                    mError = e;
+                }
             }
             mMediaRecorder.release();
         }

